File: backend/app/schemas/debug.py
import json
from datetime import datetime

from pydantic import BaseModel, field_validator


class RawRequestResponse(BaseModel):
    id: str
    turn_id: str
    provider: str
    endpoint_url: str
    request_json: dict  # Parsed JSON
    created_at: datetime

    model_config = {"from_attributes": True}

    @field_validator("request_json", mode="before")
    @classmethod
    def parse_json(cls, v):
        if isinstance(v, str):
            return json.loads(v)
        return v


class RawResponseResponse(BaseModel):
    id: str
    turn_id: str
    provider: str
    response_json: dict  # Parsed JSON
    http_status_code: int
    latency_ms: int
    token_usage_json: dict | None = None
    created_at: datetime

    model_config = {"from_attributes": True}

    @field_validator("response_json", "token_usage_json", mode="before")
    @classmethod
    def parse_json(cls, v):
        if isinstance(v, str):
            return json.loads(v)
        return v

File: frontend/src/types/debug.ts
export interface RawRequestData {
  id: string;
  turn_id: string;
  provider: string;
  endpoint_url: string;
  request_json: Record<string, unknown>;
  created_at: string;
}

export interface RawResponseData {
  id: string;
  turn_id: string;
  provider: string;
  response_json: Record<string, unknown>;
  http_status_code: number;
  latency_ms: number;
  token_usage_json: Record<string, number> | null;
  created_at: string;
}
